debian/diffoscope.1.rst
============
 diffoscope
============

-------------------------------------------------------
in-depth comparison of files, archives, and directories
-------------------------------------------------------

:Author: Debian “Reproducible Builds” Team
:Copyright: GPL-3+
:Manual section: 1
:Manual group: Debian

SYNOPSIS
========

  diffoscope [-h] [--version] [--debug] [--html output] [--text output] [--max-report-size bytes] [--css url] file1 file2

DESCRIPTION
===========

diffoscope will try to get to the bottom of what makes files or
directories different. It will recursively unpack archives of many kinds
and transform various binary formats into more human readable form to
compare them. It can compare two tarballs, ISO images, or PDF just as
easily.

It can be scripted through error codes, and a report can be produced
with the detected differences. The report can be text or HTML.
When no type of report has been selected, diffoscope defaults
to write a text report on the standard output.

diffoscope is developed as part of the “reproducible builds” Debian
project and was formerly known as “debbindiff”.

OPTIONS
=======

-h, --help               show this help message and exit
--version                show program's version number and exit
--debug                  display debug messages
--html output            write HTML report to given file
                         (use - for standard output)
--text output            write plain text report to given file
                         (use - for standard output)
--max-report-size BYTES
                         maximum bytes written in report (default: 2048000)
--max-diff-block-lines MAX_DIFF_BLOCK_LINES
                         maximum number of lines per diff block (default: 50)
--max-diff-input-lines MAX_DIFF_INPUT_LINES
                         maximum number of lines fed to diff (default: 100000)
--fuzzy-threshold FUZZY_THRESHOLD
                         threshold for fuzzy-matching (0 to disable, 60 is
                         default, 400 is high fuzziness)
--new-file               treat absent files as empty
--css url                link to an extra CSS for the HTML report

EXIT STATUS
===========

Exit status is 0 if inputs are the same, 1 if different, 2 if trouble.

SEE ALSO
========

* `<https://diffoscope.org/>`
* `<https://wiki.debian.org/ReproducibleBuilds>`


debian/changelog
diffoscope (51) unstable; urgency=medium

  * Team upload.
  * Re-upload, the last uploaded tarball was broken, missing test files.

 -- Mattia Rizzolo <mattia@debian.org>  Sun, 28 Feb 2016 11:56:35 +0000

diffoscope (50) unstable; urgency=medium

  [ Jérémy Bobbio ]
  * Remove myself from uploaders

  [ Reiner Herrmann ]
  * Show line ordering variation in deb md5sums
  * Reset mtimes of test directories and ignore ctimes (Closes: #815171)
  * Add comparator for Postscript files

 -- Holger Levsen <holger@debian.org>  Sat, 27 Feb 2016 19:37:20 +0100

diffoscope (49) unstable; urgency=medium

  [ Jérémy Bobbio ]
  * Don't crash when we can't find the debug file in a matching debug package.
    This repairs diffoscope for xserver-xorg and other packages.
  * Use libarchive to read metadata from ar archives. The output is more
    precise and less dependent on binutils version. The command line `ar` tool
    is not used any more so remove it from the required tools.
  * Split readelf --debug-dump output when examining ELF files. Based on a
    patch by Dhole.
  * Keep both .debug_str and .zdebug_str as ElfSection. Thanks to Niels Thykier
    for noticing the problem.
  * Fix a logic error in _install_debug_symbols. Thanks anthraxx for the report
    and the fix.
  * Use recursive containers for directory and only look at files with ending
    in .deb when looking for Build IDs. Both avoid looking at too many files
    when searching for matching debug packages. Thanks Helmut Grohne and Steven
    Chamberlain for the reports and tests. (Closes: #813052)
  * Add support for ICC profiles. This adds a Recommends on colord in Debian.
  * Harmonize spaces and commas in package list.
  * Update PPU files test data for FPC 3.0.0.
  * Update exepcted javap output for OpenJDK 1.8.
  * Stop specializing files from directories earlier than necessary.

  [ anthraxx ]
  * Adding cd-iccdump package reference for Arch Linux.

  [ Mattia Rizzolo ]
  * Use HTTPS in Vcs-Git.

  [ Holger Levsen ]
  * Bump standards version to 3.9.7, no changes needed.
  * Use /git/ instead /cgit/ in Vcs-Browser.

 -- Jérémy Bobbio <lunar@debian.org>  Wed, 17 Feb 2016 18:12:45 +0100

diffoscope (48) unstable; urgency=medium

  * Open debian/control file in binary mode (Closes: #812524)
  * Fix recognizes() for files ending in .buildinfo but without a
    Checksums-Sha256 field. (Closes: #812534)
  * Fix finding debug packages with multiple Build-Ids.
  * Cleanup how arguments are given to readelf for sections.
  * Only pass --decompress to readelf when the option is supported. This
    restores compatibility with older versions of binutils.
  * Skip dbgsym tests if the debian Python module is not available.

 -- Jérémy Bobbio <lunar@debian.org>  Tue, 26 Jan 2016 12:13:57 +0100

diffoscope (47) unstable; urgency=low

  * Don't show a difference for identical but long diff inputs (e.g. objdump
    output). This should make diffoscope run longer for large diff inputs, but
    more correct reports worths it.
  * Fix symbol ignore regexp for ELF files.
  * Ask readelf to decompress sections before dumping their content.
  * Don't redefine .name() in Archive when its already defined in File.
  * All files can now belong to a container. This will help us implement
    cross-file analysis.
  * Update requirements for ELF tests.
  * Sort file order when comparing directories.
  * Use Python OrderedDict instead of keeping a list of sections in ELF
    container.
  * Install detached symbols from debug .deb before comparing ELF files.
    This means objdump output should have line numbers for Debian packages
    built with recent debhelper as long as the associated debug package is in
    the same directory.
  * Add support for Debian .buildinfo files.

 -- Jérémy Bobbio <lunar@debian.org>  Sat, 23 Jan 2016 18:08:33 +0000

diffoscope (46) unstable; urgency=low

  [ Dhole ]
  * Improve ELF comparisons by performing comparisons section by section.
    (Closes: #808197)

  [ Jérémy Bobbio ]
  * Further split readelf output when examining ELF files.
  * Skip dumping ELF sections that are already covered by previous readelf
    calls to avoid redundant output. Thanks Mike Hommey for the report.
    (Closes: #808267)
  * Fix ELF comparisons against non-existing files.
  * Filter symbols for IP-relative ops from ELF disassembled output as
    they will create differences that don't really exist. Thanks Mike Hommey
    for the report and test cases. (Closes: #808207)
  * Add forgotten requirements for tests with non-existing files.
  * Improve tests for comparisons against non-existing files.
  * Respect order of containers when performing comparisons. It makes the
    output somewhat nicer, especially for things that are expected to be in a
    certain order, like ELF sections.
  * Fix comparisons of directory with broken symlinks. Thanks Tuomas Tynkkynen
    for reporting the issue. (Closes: #810825)

 -- Jérémy Bobbio <lunar@debian.org>  Tue, 19 Jan 2016 16:46:47 +0100

diffoscope (45) unstable; urgency=medium

  [ Rainer Müller ]
  * Fix typo in Mach-O comparator.

  [ Jérémy Bobbio ]
  * Ensure that we always get path names from libarchive as str.
    (Closes: #808541)
  * Read ext4 reference diff using UTF-8 to ensure that the test suite
    works when run under various locales.
  * Replace file path by a fixed placeholder in readelf and objdump output.
    This will work better when comparing ELF files given on the command line.
  * Read lines using an iterator instead of loading a full list in memory.
    This greatly help diffoscope when running on large binaries. Thanks Mike
    Hommey for the report and good test case. (Closes: #808120)
  * Properly report lines we can't parse in squashfs listing.
  * Correctly parse squashfs device entries with device minors larger than 3
    digits. Thanks sajolida for providing the samples.
  * Use libarchive to list files in tar and cpio archives. The output of cpio

diffoscope.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# diffoscope: in-depth comparison of files, archives, and directories
#
# Copyright © 2014-2015 Jérémy Bobbio <lunar@debian.org>
#
# diffoscope is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# diffoscope is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with diffoscope.  If not, see <http://www.gnu.org/licenses/>.

from __future__ import print_function

import argparse
from contextlib import contextmanager
import logging
import codecs
import os
import sys
import traceback
from diffoscope import logger, VERSION, set_locale
import diffoscope.comparators
from diffoscope.config import Config
from diffoscope.presenters.html import output_html
from diffoscope.presenters.text import output_text


def create_parser():
    parser = argparse.ArgumentParser(
        description='Highlight differences between two builds '
                    'of Debian packages')
    parser.add_argument('--version', action='version',
                        version='diffoscope %s' % VERSION)
    parser.add_argument('--list-tools', nargs=0, action=ListToolsAction,
                        help='show external tools required and exit')
    parser.add_argument('--debug', dest='debug', action='store_true',
                        default=False, help='display debug messages')
    parser.add_argument('--debugger', action='store_true',
                        help='Open the python debugger in case of crashes.')
    parser.add_argument('--html', metavar='output', dest='html_output',
                        help='write HTML report to given file (use - for stdout)')
    parser.add_argument('--text', metavar='output', dest='text_output',
                        help='write plain text output to given file (use - for stdout)')
    parser.add_argument('--max-report-size', metavar='BYTES',
                        dest='max_report_size', type=int,
                        help='maximum bytes written in report')
    parser.add_argument('--max-diff-block-lines', dest='max_diff_block_lines', type=int,
                        help='maximum number of lines per diff block')
    parser.add_argument('--max-diff-input-lines', dest='max_diff_input_lines', type=int,
                        help='maximum number of lines fed to diff')
    parser.add_argument('--css', metavar='url', dest='css_url',
                        help='link to an extra CSS for the HTML report')
    parser.add_argument('file1', help='first file to compare')
    parser.add_argument('file2', help='second file to compare')
    return parser


@contextmanager
def make_printer(path):
    if path == '-':
        output = sys.stdout
        if not sys.stdout.isatty():
            output = codecs.getwriter(sys.stdin.encoding)(sys.stdout)
    else:
        output = codecs.open(path, 'w', encoding='utf-8')
    def print_func(*args, **kwargs):
        kwargs['file'] = output
        print(*args, **kwargs)
    yield print_func
    if path != '-':
        output.close()


class ListToolsAction(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        from diffoscope import tool_required, RequiredToolNotFound
        print("External tools required:")
        print(', '.join(tool_required.all))
        print()
        print("Available in packages:")
        print(', '.join(sorted(set([RequiredToolNotFound.PROVIDERS[k]["debian"] for k in tool_required.all]))))
        sys.exit(0)


def main():
    Config.general.max_diff_block_lines = parsed_args.max_diff_block_lines
    Config.general.max_diff_input_lines = parsed_args.max_diff_input_lines
    Config.general.max_report_size = parsed_args.max_report_size
    if parsed_args.debug:
        logger.setLevel(logging.DEBUG)
    set_locale()
    difference = diffoscope.comparators.compare_root_paths(
        parsed_args.file1, parsed_args.file2)
    if difference:
        if parsed_args.html_output:
            with make_printer(parsed_args.html_output) as print_func:
                output_html(difference, css_url=parsed_args.css_url, print_func=print_func)
        if (parsed_args.text_output and parsed_args.text_output != parsed_args.html_output) or not parsed_args.html_output:
            with make_printer(parsed_args.text_output or '-') as print_func:
                output_text(difference, print_func=print_func)
        return 1
    return 0

if __name__ == '__main__':
    try:
        parser = create_parser()
        parsed_args = parser.parse_args(sys.argv[1:])
        sys.exit(main())
    except KeyboardInterrupt:
        logger.info('Keyboard Interrupt')
        sys.exit(2)
    except Exception:
        traceback.print_exc()
        if parsed_args.debugger:
            import pdb
            pdb.post_mortem()
        sys.exit(2)
